### PR TITLE
Feature/reference set

### PR DESCRIPTION
**Pull request by Stanley-Okwii**
- Add reference set functionality

### CODE REVIEW

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ export class DropdownTypeahead extends Component<DropdownTypeaheadProps> {
     }
 
     private createSelectorProp(): { placeholder?: string, value?: ReferenceOption | null } {
-        if (this.props.selectedValue) {
+        if (JSON.stringify(this.props.selectedValue) !== "") {
```
Comment (jseremba) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: change to `if(JSON.stringify(this.props.selectedValue))`

**Review by jseremba — CHANGES_REQUESTED** [review]

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 2]
```diff
@@ -78,6 +78,13 @@ export class DropdownTypeahead extends Component<DropdownTypeaheadProps> {
                     className: classNames("widget-dropdown-type-ahead-wrapper")
                 },
                 this.props.selectType === "normal" ?
+                    this.props.isReadOnly ?
```
Comment (jseremba) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: I think using an if statement for such a ternary operator is far better 
`if(this.props.selectType === "normal" && this.props.isReadOnly`

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 3]
```diff
@@ -93,16 +100,38 @@ export class DropdownTypeahead extends Component<DropdownTypeaheadProps> {
...
+                if (this.props.data) {
+                    this.props.data.forEach((dataObject: any) => {
+                        const value = dataObject.value;
+                        if (value === selectedGuid) {
+                            selectedLabel = dataObject.label;
+                        }
+                    });
+                }
+
+                return selectedLabel !== "" ? selectedLabel : undefined;
```
Comment (jseremba) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: change to ` return selectedLabel || undefined;`

**Review by jseremba — CHANGES_REQUESTED** [review]

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 3 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 1 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 2 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: if condition cant not allowed within the render method

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+export const validateProps = (props: ContainerProps): string => {
+    const message: string[] = [];
+
+    if (props.onChangeEvent === "callMicroflow" && !props.onChangeMicroflow) {
+        message.push("On change event is set to 'Call a microflow' but no microflow is selected");
+    } else if (props.onChangeEvent === "callNanoflow" && !props.onChangeNanoflow.nanoflow) {
+        message.push("On change event is set to 'Call a nanoflow' but no nanoflow is selected");
+    }
+
+    if (props.labelCaption.trim() === "" && props.showLabel && (props.labelWidth > 11 || props.labelWidth < 1)) {
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: remove check for empty captio
```suggestion
    if (props.showLabel && (props.labelWidth > 11 || props.labelWidth < 1)) {
```

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        message.push("Label width should be a value between 0 and 12");
+    }
+
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: Missing check for required datasource Microflow, and nanoflow

File: src/DropdownTypeaheadReferenceSet/components/Label.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    const labelWeight = orientation === "horizontal" ? `col-sm-${weight}` : "";
+    const childrenWeight = orientation === "horizontal" ? `col-sm-${12 - weight}` : "";
+
+    return createElement("div", { className: classNames("form-group", className), style },
+        createElement("label", { className: `control-label ${labelWeight}` }, label),
+        createElement("div", { className: `${childrenWeight}` }, children)
+    );
+};
+
+Label.defaultProps = { weight: 6 };
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/Label.ts]: form default is 3
```suggestion
Label.defaultProps = { weight: 3 };
```

**Review by Andries-Smit — COMMENTED** [review]
Continue review on inital PR

File: src/DropdownTypeahead/components/DropdownTypeaheadContainer.ts [hunk 7]
```diff
@@ -1,9 +1,9 @@
-import { ChangeEvent, Component, createElement } from "react";
+import { Component, createElement } from "react";
 import * as initializeReactFastclick from "react-fastclick";
```
Comment (Andries-Smit) [inline src/DropdownTypeahead/components/DropdownTypeaheadContainer.ts]: Is fast click still needed, or was this for version 3 of the dropdown lib

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        fetchData(options)
+            .then(optionObjects => this.setOptions(optionObjects))
+            .catch(mx.ui.error);
+    }
+
+    private setOptions = (mendixObjects: mendix.lib.MxObject[]) => {
+        const options: ReferenceOption[] = [];
+
+        if (mendixObjects.length > 0) {
+            mendixObjects.forEach(mxObject => {
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: No need to do a length check before `forEach`

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            mxform: this.props.mxform,
+            nanoflow,
+            sortAttributes: this.props.sortAttributes,
+            sortOrder,
+            source
+        };
+
+        fetchData(options)
+            .then(optionObjects => this.setOptions(optionObjects))
+            .catch(mx.ui.error);
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: Probably not working when catching an error in a microflow

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+export interface AttributeType {
+    name: string;
+    sort: string;
+}
+
+export class DropdownTypeahead extends Component<DropdownTypeaheadProps> {
+    render() {
+        return this.renderForm();
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: This extracted function does not add any value, as `renderForm` is only referenced once

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                    orientation: this.props.labelOrientation,
+                    style: this.props.styleObject,
+                    weight: this.props.labelWidth
+                }, this.renderSelector());
+            }
+
+            return this.renderSelector();
+        }
+
+        return createElement("div", {});
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: a) Strange to render a div without a class.
b) Options are optional `createElement("div");`
c) Strange to not show anything...

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            alertMessage: validateProps(this.props),
+            className: this.props.class,
+            data: this.state.options,
+            asyncData: this.setAsyncOptions,
+            emptyOptionCaption: this.props.emptyOptionCaption,
+            handleOnchange: this.handleOnClick,
+            isClearable: this.props.isClearable,
+            selectType: this.props.selectType,
+            isReadOnly: this.isReadOnly(),
+            loaded: this.state.isLoading,
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: dont onderstand... they are not interchangeable

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    sort: string;
+}
+
+export class DropdownTypeahead extends Component<DropdownTypeaheadProps> {
+    render() {
+        return this.renderForm();
+    }
+
+    private renderForm() {
+        if (!this.props.loaded) {
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: or should the prob be named `loading...`
```suggestion
        if (this.props.loaded) {
```

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+            if (this.props.showLabel && this.props.labelCaption) {
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: no need to check for caption.
```suggestion
            if (this.props.showLabel) {
```

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    if (props.labelCaption.trim() === "" && props.showLabel) {
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: Remove check, we should support empty captions too. Just like the core widgets

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+                    this.props.isReadOnly ?
+                    createElement("input", {
+                        type: "text",
+                        readonly: "readonly",
+                        className: "form-control",
+                        disabled: "disabled",
+                        value: this.processOptions() }) :
+                    createElement(Select, {
+                        options: this.props.data,
+                        noResultsText: "No options",
```
Comment (Andries-Smit) [inline src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Should be support this? if so it should be translatable String
For now make text empty
```suggestion
                        noResultsText: "",
```

**Review by Andries-Smit — CHANGES_REQUESTED** [review]
Part 2

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts (hunk 4 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: Done

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts (hunk 5 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: Added

File: src/DropdownTypeahead/components/DropdownTypeaheadContainer.ts (hunk 7 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeahead/components/DropdownTypeaheadContainer.ts]: Yes, it is until I have a better a solution or work around it.

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts (hunk 8 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 16 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Done

File: src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts (hunk 15 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/utils/ContainerUtils.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/Label.ts (hunk 6 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/Label.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 14 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 13 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Done

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 10 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: Removed it

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts (hunk 12 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded' to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: I have used 'loaded'  as state to be uniform and consistent

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts (hunk 9 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSetContainer.ts]: Fixed it

File: src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts (hunk 11 above)
Comment (Stanley-Okwii) [reply src/DropdownTypeaheadReferenceSet/components/DropdownTypeaheadReferenceSet.ts]: I have included a 'spinner' to show users that data is being retrieved

**Comment by Andries-Smit** [comment]
PR not relevant to wrong branche

**Pull request closed without merging**